Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs

```diff
@@ -92,7 +121,105 @@ public void Dispose()
             RetryPolicy retryPolicy,
             CancellationToken cancellationToken = default)
         {
-            return await this.sessionPool.Execute<T>(func, retryPolicy, cancellationToken);
+            if (this.isClosed)
+            {
+                this.logger.LogError(ExceptionMessages.DriverClosed);
+                throw new QldbDriverException(ExceptionMessages.DriverClosed);
+            }
+
+            bool replaceDeadSession = false;
+            int retryAttempt = 0;
+            while (true)
+            {
+                AsyncQldbSession session = null;
+                try
+                {
+                    if (replaceDeadSession)
+                    {
+                        session = await this.StartNewSession();
+                    }
+                    else
+                    {
+                        session = await this.GetSession();
+                    }
+
+                    T returnedValue = await session.Execute(func);
+                    this.ReleaseSession(session);
+                    return returnedValue;
+                }
+                catch (RetriableException re)
+                {
+                    // If initial session is invalid, always retry once with a new session.
+                    if (re.InnerException is InvalidSessionException && retryAttempt == 0)
+                    {
+                        this.logger.LogDebug("Initial session received from pool invalid. Retrying...");
+                        replaceDeadSession = true;
+                        retryAttempt++;
+                        continue;
+                    }
+
+                    // Normal retry logic.
+                    if (retryAttempt++ >= retryPolicy.MaxRetries)
+                    {
+                        if (re.IsSessionAlive)
+                        {
+                            this.ReleaseSession(session);
+                        }
+                        else
+                        {
+                            this.poolPermits.Release();
+                        }
+
+                        throw re.InnerException;
+                    }
+
+                    this.logger.LogInformation("A recoverable error has occurred. Attempting retry #{}.", retryAttempt);
+                    this.logger.LogDebug(
+                        "Errored Transaction ID: {}. Error cause: {}",
+                        re.TransactionId,
+                        re.InnerException.ToString());
+                    replaceDeadSession = !re.IsSessionAlive;
```

Comment:
Why on Java is this about abort, but here's it's about being alive?

Reply:
Meaning is the same, but I'll update it on the Java side since I think this makes more sense for the context it's used in.